lib/CodeGen/CGObjCMac.cpp

```diff
@@ -7247,19 +7266,38 @@ CGObjCNonFragileABIMac::EmitClassRefFromId(CodeGenFunction &CGF,
     llvm::Constant *ClassGV;
     if (ID) {
       ClassGV = GetClassGlobal(ID, /*metaclass*/ false, NotForDefinition);
+
+      if (ID->hasAttr<ObjCClassStubAttr>()) {
+        ClassGV = llvm::ConstantExpr::getPtrToInt(ClassGV, CGM.IntPtrTy);
+
+        // Classrefs pointing at Objective-C stub classes have the least
+        // significant bit set to 1.
+        auto *Tag = llvm::ConstantInt::get(CGM.IntPtrTy, 1);
+        ClassGV = llvm::ConstantExpr::getAdd(ClassGV, Tag);
+        ClassGV = llvm::ConstantExpr::getIntToPtr(ClassGV,
+                                                  ObjCTypes.Int8PtrTy);
+      }
     } else {
       ClassGV = GetClassGlobal((getClassSymbolPrefix() + II->getName()).str(),
                                NotForDefinition);
     }
 
-    Entry = new llvm::GlobalVariable(CGM.getModule(), ObjCTypes.ClassnfABIPtrTy,
```

Comment:
What happens to this type? Can we assert that it's the correct type, at least?

Reply:
It's going to be Int8PtrTy in one case and ClassnfABIPtrTy in the other. I could pretend its a ClassnfABIPtrTy in both, but in the stub case the pointed-to-thing is not a Class, and the pointer is offset by 1.

Reply:
Yes, I see that, but in the old code we checked that we had a ClassnfABIPtrTy. Is it important to preserve that check in the normal case?

Reply:
I added the assert. The Clang CL for this patch is here, you can post review comments there instead: https://reviews.llvm.org/D59628